funnel/transports/sms.py

```diff
@@ -9,65 +17,131 @@
-                sid=sid
-            ),
-            auth=(sid, token),
-            data={
-                'From': app.config['SMS_TWILIO_FROM'],
-                'To': phone_number,
-                'Body': message,
-            },
+__all__ = ['send', 'SmsSender', 'TwilioSmsResponse']
+
+
+@dataclass_json
+@dataclass
+class TwilioSmsResponse:
+    """
+    SMS Status Message sent by Twilio for processing. See this for more information
+    https://www.twilio.com/docs/sms/send-messages
+
+    * sms_sid: Response ID of the SMS.
+    * status:  These could be any of queued, failed, sent, delivered, undelivered
+    * msg_status: Same values as status
+    * to: Destination Phone number
+    * msg_sid: Same values as sms_sid
+    * account_sid: Sending Account ID
+    * sender: Sending Phone number
+    * api_version: API Version
+    """
+
+    sms_sid: str = field(metadata=config(field_name='SmsSid'))
+    status: str = field(metadata=config(field_name='SmsStatus'))
+    msg_status: Optional[str] = field(metadata=config(field_name='MessageStatus'))
+    to: str = field(metadata=config(field_name='To'))
+    msg_sid: Optional[str] = field(metadata=config(field_name='MessageSid'))
+    account_sid: str = field(metadata=config(field_name='AccountSid'))
+    sender: str = field(metadata=config(field_name='From'))
+    api_version: Optional[str] = field(metadata=config(field_name='ApiVersion'))
+
+
+class SmsSender:
+    """ Manages Sending SMS to different providers. """
+
+    def __init__(self, callback: bool = True):
+        self.twilio_callback = None
+        self.exotel_callback = None
+        self.callback = callback
+
+    def send(self, phone_number: str, message: str) -> str:
+        """
+        Send the message to a given phone number based on internal rules.
+        :param phone_number: Phone Number
+        :param message: Message to deliver to Phone number.
+        :return SMS Message ID
+        """
+        if phone_number.startswith('+91'):
+            return self._route_via_exotel(phone_number, message)
+        elif phone_number.startswith('+'):
+            return self._route_via_twilio(phone_number, message)
+        raise TransportRecipientError(
+            "No service provider available for this recipient"
         )
-        if r.status_code in (200, 201):
-            # All good
-            jsonresponse = r.json()
-            return jsonresponse.get('sid')
-        raise TransportTransactionError("Twilio API error", r.status_code, r.text)
-    except requests.ConnectionError:
-        raise TransportConnectionError("Twilio not reachable")
+
+    @staticmethod
+    def _route_via_exotel(phone_number: str, message: str) -> str:
+        """
+        Route the SMS via Twilio. This is done only for messages that can't be delivered by exotel.
+        :param phone_number: Phone Number
+        :param message: Message to deliver to Phone number.
```

Comment:
Sphinx (documentation generator) supports type annotations in `param` tags. Syntax: `:param type name: doc`

We don't use Sphinx to document Funnel, but we should. Also, this _may_ be unnecessary if Sphinx reads type annotations from the function signature itself. I haven't checked for this since moving to Py3.

Reply:
Confirmed, Sphinx gets it from annotations. This fix is not necessary. https://stackoverflow.com/questions/34647966/how-to-express-multiple-types-for-a-single-parameter-or-a-return-value-in-docstr